### DISCUSSION TITLE
Custom role permissions for secrets and envs?

### DISCUSSION THREAD

**Discussion by fishpen0** [community/community#40940] [Actions]
### Select Topic Area

Product Feedback

### Body

How do I give devs access to manage github actions secrets and environments without making them a respository admin?  I am aware of custom roles, but looking through the role settings, envs and secrets appear to not be mapped to any permissions available to select from?

Also, there is still a mismatch between UI access and API access for some of these permissions, for example https://github.com/github/docs/issues/1087 is still an issue

**Comment by xv-carlos-c** [comment]
2 years after and this is still not possible

**Comment by mijdavis2** [comment]
User's with `write` and `maintain` having the ability to create, update, and delete repo secrets via the API but NOT in the UI is extremely confusing and forces us to give full admin permissions for those that need to use the console instead of the api, which gives them undesired permissions as well. Please make this permission available in custom roles or make the UI perms consistent with the API perms.

This is incredibly unexpected and undesirable
![image](https://github.com/user-attachments/assets/9027f87a-a99d-4b45-848c-3ac891ad1e99)

**Reply by rajbos** [reply]
For people stumbling upon this, I documented things here: https://devopsjournal.io/blog/2022/11/02/GitHub-secrets-without-admin-rights

**Comment by Nigmatullayev001** [comment]
Answer:
You don't need to grant repository admin access to allow developers to manage GitHub Actions secrets and environments. Here are some alternative approaches:

🔹 1. Using Custom Roles (GitHub Enterprise)
If you are using GitHub Enterprise, you can create Custom Roles and assign specific permissions. However, as you mentioned, secrets and environments do not currently map directly to any selectable permissions, which is a limitation.

🔹 2. Using Environment-Based Secrets
Instead of using repository-level secrets, you can configure environment-based secrets, which offer more granular control:

Go to Settings → Environments → Create a new Environment.
Enable "Required reviewers" to control access.
Assign permissions only to specific users or teams.
This way, developers can access secrets only for specific environments without being repository admins.

🔹 3. Managing Permissions Without Admin Access
If you want to allow users to work with Actions without making them admins, you can grant them:

Write access → Allows triggering workflows.
Maintain access → Allows managing GitHub Actions without full admin control.
These permissions can be managed under Repository Settings → Collaborators & Teams.

🔹 4. UI vs. API Permission Mismatch (Bug)
You're right that github/docs#1087 is still an issue. Some permissions work via API but are not yet fully supported in the UI. This is an ongoing limitation in GitHub's permission model.

📌 Workaround: If API restrictions are affecting your workflow, I’d recommend submitting feedback to GitHub Support or discussing it in GitHub Discussions to raise awareness.

I hope this helps! 😊 🚀

**Reply by mijdavis2** [reply]
This is clearly an AI spam response, WTH?! This was not helpful at all as all your points are completely invalid for this use case.
